GEODE-8864:finish implementation of Redis HScan Command

jhutchison opened the pull request:
-add test for hscan not returning value removed from hash

-refactor some tests to better reflect use cases

-create HScan DunitTest

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

sabbey37 commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -60,34 +65,56 @@ public void tearDown() {
+
+  @Test
+
+  public void givenMatchArgumentWithoutPatternOnNonExistentKey_returnsEmptyArray() {
+    jedis.hset("key", "b", "1");
+
+    ScanParams scanParams = new ScanParams();
+    scanParams.match("");
+
+    ScanResult<Map.Entry<String, String>> result =
+        jedis.hscan("key", "0", scanParams);
+
+    assertThat(result.getResult()).isEmpty();
+  }
```
The name of this test seems inaccurate.  It says it is a `NonExistentKey` however, we do an hset beforehand with the key, so it should exist with data in the cache. Also, we say we don't provide a pattern, but empty string actually is a pattern.  What we could do (if we want to test a match argument without a pattern on a nonexistent key), is not do the hset, then do:
```
List<Object> result =
        (List<Object>)  jedis.sendCommand(Protocol.Command.HSCAN, "key1", "0", "Match");

assertThat((List<String>) result.get(1)).isEmpty();
```

jhutchison replied:
thanks-  will revert to original send command

jhutchison replied:
and thanks for catching the name

sabbey37 commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -60,34 +65,56 @@ public void tearDown() {
   @SuppressWarnings("unchecked")
-  public void givenArgumentsAreNotOddAndKeyDoesNotExist_returnsEmptyArray() {
+  public void givenCountArgumentWithoutNumberOnNonExistentKey_returnsEmptyArray() {
+    jedis.hset("a", "b", "1");
```
We don't need to do an hset here.

jhutchison replied:
true.  will take out

sabbey37 commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -121,17 +156,10 @@ public void givenCount_whenCountParameterIsNegative_returnsSyntaxError() {
-        "COUNT", "sjlfs", "COUNT", "1"))
-            .hasMessageContaining(ERROR_NOT_INTEGER);
-  }
+  public void givenMultipleCounts_whenAnyCountParameterIsLessThanOne_DoesNotError() {
+    jedis.hset("key", "b", "1");
```
The title of this test seems incorrect.  When any count parameter is less than 0, we return a syntax error.  maybe instead of saying `DoesNotError` we could change it back to `returnsSyntaxError`

jhutchison replied:
will do

sabbey37 reviewed (changes requested):
I like the reorganization of the integration tests!  Just a few recommended changes.  If we are making this a supported command, it needs to be updated in `RedisCommandType` as well as the `SupportedCommandsJUnitTest`

jdeppe-pivotal commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,178 @@
+        (i) -> multipleHScanLamba(allEntries2, jedis2),
+        (i) -> jedis3.hset(HASH_KEY, BASE_FIELD + i, "new_value_" + i)).run();
+
+    assertThat(allEntries1.size())
```
I'm not sure this is quite what you want. Since `multipleHScanLamba` always clears the entries, this assert is only ever checking the result of the last iteration and not all of them. It might be better it simply count the results in the lambda and then assert on `ITERATION_COUNT * SIZE_OF_INITIAL_HASH_DATA`

jdeppe-pivotal commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,178 @@
+          jedis3.hdel(HASH_KEY, field);
+        }).run();
+
+    assertThat(allEntries1).containsOnlyOnceElementsOf(INITIAL_HASH_DATA.entrySet());
```
Similar problem to the above.

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -121,17 +156,10 @@ public void givenCount_whenCountParameterIsNegative_returnsSyntaxError() {
-        "COUNT", "sjlfs", "COUNT", "1"))
-            .hasMessageContaining(ERROR_NOT_INTEGER);
-  }
+  public void givenMultipleCounts_whenAnyCountParameterIsLessThanOne_DoesNotError() {
```
will do

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +418,31 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
+    entryMap.put("3", "grey");
     jedis.hmset("colors", entryMap);
+    Future deleteFuture = doAsyncHdel("colors", "3");
+    deleteFuture.get();
```
Yeah, I see your point.  Let me look at what I thought I was doing here...

jhutchison replied:
so yeah, the point was actually to make sure that the hdel WAS synchronous (had completed before the hscan started) .   Switched to use a geode awaitutility, I guess that is a little more straightforward.  The test itself seems a little obvious, I agree, but it's called out in the docs, so In figured it was good to have a test

jdeppe-pivotal commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,241 @@
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class HScanDunitTest {
```
I think these would be very valuable to run as integration tests so that we can validate native redis' functionality.

jhutchison replied:
will move them

jdeppe-pivotal commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,241 @@
+        (i) -> multipleHScanAndAssertOnSizeOfResultSet(jedis1),
+        (i) -> multipleHScanAndAssertOnSizeOfResultSet(jedis2),
+        (i) -> {
+          int fieldSuffix =
```
This can simply just be `i % SIZE_OF_INITIAL_HASH_DATA`

jhutchison replied:
thanks-  that's simpler

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -45,10 +47,14 @@
 public class RedisHash extends AbstractRedisData {
   public static final RedisHash NULL_REDIS_HASH = new NullRedisHash();
   private HashMap<ByteArrayWrapper, ByteArrayWrapper> hash;
+  private HashMap<UUID, Map<ByteArrayWrapper,ByteArrayWrapper>> perClientHScanDataSnapShot;
```
We should only be storing the list of keys here and not the whole hash. Redis doesn't guarantee the _value_ of a field remaining constant across scans just the list of fields. It's bad enough that we need to keep a list of fields but would be even worse if we also need to keep the values.

jhutchison replied:
ok-  changed to just store the keys

nonbinaryprogrammer commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,142 @@ public int hstrlen(ByteArrayWrapper field) {
+      int numberOfIterationsCompleted) {
+
+    if (numberOfIterationsCompleted < count) {
+      return BigInteger.valueOf(0);
```
why not return an integer here that the caller can then create a BigInteger from if that's what they need?

jhutchison replied:
done

nonbinaryprogrammer reviewed (changes requested):
when you do cleanup please check the diff for unnecessary line breaks. There is one file that only has formatting changes that I don't think you meant to commit. I also think that there are a lot of copies going on, especially in the hscan executor that should be revisited to see if any of those can be removed for the sake of garbage collection.
Overall this is looking good, sorry to inundate you with comments.

jhutchison commented on geode-redis/src/commonTest/java/org/apache/geode/redis/ConcurrentLoopingThreads.java:
```diff
@@ -63,13 +65,21 @@ private ConcurrentLoopingThreads start(boolean lockstep) {
-        loopingThread.get();
-      } catch (InterruptedException | ExecutionException e) {
-        throw new RuntimeException(e);
+    boolean retry = false;
```
done

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/RedisHashCommandsFunctionInvoker.java:
```diff
@@ -106,9 +107,10 @@ public int hstrlen(ByteArrayWrapper key, ByteArrayWrapper field) {
+      Pattern matchPattern, int count,
       BigInteger cursor) {
-    return invokeCommandFunction(key, HSCAN, matchPattern, count, cursor);
+    return invokeCommandFunction(key, HSCAN, matchPattern, count, cursor, clientID);
```
good catch-  thx

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/netty/Client.java:
```diff
@@ -26,8 +27,15 @@
 public class Client {
   private Channel channel;
 
+  public UUID getId() {
+    return id;
+  }
```
done

jhutchison commented on geode-redis/src/test/java/org/apache/geode/redis/internal/data/RedisHashTest.java:
```diff
@@ -161,4 +163,79 @@ public void setExpirationTimestamp_stores_delta_that_is_stable() throws IOExcept
+
+    UUID clientID = UUID.randomUUID();
+
+    subject.hscan(clientID, null, 10, BigInteger.valueOf(0));
```
added

jhutchison commented on geode-redis/src/test/java/org/apache/geode/redis/internal/data/RedisHashTest.java:
```diff
@@ -161,4 +163,79 @@ public void setExpirationTimestamp_stores_delta_that_is_stable() throws IOExcept
+    List<ByteArrayWrapper> keyList = hscanSnapShotMap.get(clientID);
+    assertThat(keyList).isNotEmpty();
+
+    assertThat(hscanSnapShotMap.get(clientID).contains("field_1"));
+    assertThat(hscanSnapShotMap.get(clientID).contains("value_1"));
+    assertThat(hscanSnapShotMap.get(clientID).contains("field_8"));
+    assertThat(hscanSnapShotMap.get(clientID).contains("value_8"));
```
chnaged

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +453,120 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
-    String cursor = "-100";
-    ScanResult<Map.Entry<String, String>> result;
-    List<Map.Entry<String, String>> allEntries = new ArrayList<>();
+    Map<String, String> originalData = new HashMap<>();
```
done

jhutchison replied:
done

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +453,120 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
+        (i) -> {
+          int fieldSuffix = i % SIZE_OF_INITIAL_HASH_DATA;
+          jedis3.hset(HASH_KEY, BASE_FIELD + fieldSuffix, "new_value_" + i);
+        }).run();
```
added a new test that checks that data is not altered by concurrent hscans

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +453,120 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
+          String field = "new_" + BASE_FIELD + i;
+          jedis3.hset(HASH_KEY, field, "whatever");
+          jedis3.hdel(HASH_KEY, field);
+        }).run();
```
added a new test that checks that data is not altered by concurrent hscans

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,142 @@ public int hstrlen(ByteArrayWrapper field) {
-          returnList.add(value);
-          numElements++;
-        }
+        addMatches(matchPattern, resultList, key, value);
```
done

jdeppe-pivotal commented on geode-redis/README.md:
```diff
@@ -167,53 +167,53 @@ start server \
-|                    	|                                             	        | COMMAND COUNT                 	|
-|                    	|                                                    	| COMMAND GETKEYS               	|
-|                    	|                                                    	| COMMAND INFO                  	|
+| HSCAN                 | GETRANGE                                              | ACL USERS                     	|    
```
We were going to make all doc-related changes as one separate commit, so this can be reverted.

jhutchison replied:
ok, rolled back

jdeppe-pivotal commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -219,12 +301,13 @@ public void givenCount_returnsAllEntriesWithoutDuplicates() {
+    assertThat(allEntries.size()).isCloseTo(3, Offset.offset(1));
   }
 
+  @Ignore
```
Unless there is functionality here that will eventually get fixed, this test should just be removed. It might be better to simply assert that it is not an error to provide multiple `COUNT` arguments.

jhutchison replied:
good point-  removed the assert on the number fo values returned

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +52,71 @@
+  private HashMap<UUID, List<ByteArrayWrapper>> hScanSnapShots;
+  private HashMap<UUID, Long> hScanSnapShotCreationTimes;
+  private ScheduledExecutorService HSCANSnapshotExpirationExecutor = null;
+  private int HSCAN_SNAPSHOTS_EXPIRE_MILLISECONDS;
+  private final int TIME_FOR_SNAPSHOTS_TO_LIVE;
```
I think these fields are redundant. This value should also be configurable by the user somehow. Please could you raise a Jira for that.

jhutchison replied:
renamed these to 
HSCAN_SNAPSHOTS_EXPIRE_CHECK_FREQUENCY_MILLISECONDS
MINIMUM_MILLISECONDS_FOR_HSCAN_SNAPSHOTS_TO_LIVE 
to clarify intent.  I could still be convinced that they could be the same value, but I feel like there might be cases where users want to have infrequent checks, but want the snapshots to expire soon after creation? If you still wan them combined I won't argue.  I'll put in the ticket to make them configureable

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +52,71 @@
+
+  private void expireHScanSnapshots() {
+
+    this.hScanSnapShotCreationTimes.entrySet().forEach(entry -> {
```
There's a very small window for a `ConcurrentModificationException` to happen if something else is modifying this collection at the same time. Please raise a Jira to fix in the future.

jhutchison replied:
I couldn't decide if the length a snapshot should stick around should be different that how often to check if any need to be expired.  I could be persuaded that they are the same thing.  I can see situation where a user would want to minimize the frequency of the checks, but not want them to stay around very long (or vice versa).  for now I renamed them to  HSCAN_SNAPSHOTS_EXPIRE_CHECK_FREQUENCY_MILLISECONDS and  MINIMUM_MILLISECONDS_FOR_HSCAN_SNAPSHOTS_TO_LIVE to make the intent more clear.  If you feel strongly there should only be one let me know.    I'll add the jira.

jhutchison replied:
as per convo, changed hscanpSnaphot and creationtimes to concurrent hashmaps

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,120 @@ public int hstrlen(ByteArrayWrapper field) {
+
+    for (ByteArrayWrapper key : keysSnapShot) {
+
+      if (indexOfKeys < startCursor) {
+        indexOfKeys++;
+        continue;
+      }
+
+      indexOfKeys++;
```
Since `keysSnapShot` is indexable, can this be a simple `for` loop that starts at `startCursor` and avoid this 'skipping' logic entirely?

jhutchison replied:
yeah, much better-  thanks

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,120 @@ public int hstrlen(ByteArrayWrapper field) {
+    List<Object> resultList =
+        getResultList(keysToScan, startCursor, count, matchPattern);
+
+    int indexOfLast = resultList.size() - 1;
+    int numberOfIterationsCompleted = Integer.parseInt(resultList.get(indexOfLast).toString());
+    resultList.remove(indexOfLast);
```
See below: rather return a `Pair` so that the intent is clearer.

jhutchison replied:
done.  yeah, I really like having to return another another type here, but I don't see a simple way around it. open to suggestions...

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,120 @@ public int hstrlen(ByteArrayWrapper field) {
       }
 
-      if (numElements == count) {
+      if ((indexOfKeys - startCursor) == count) {
```
Should this not instead be checking the size of `resultList` so that when you've actually accumulated `count` values it will return?

jhutchison replied:
As per Redis docs, the match is applied after the count of elements are received.  We could do it differently...

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +261,120 @@ public int hstrlen(ByteArrayWrapper field) {
       }
+
+    }
+    resultList.add(new ByteArrayWrapper(String.valueOf(indexOfKeys - startCursor).getBytes()));
```
If you do need to return 2 distinct values, please use a `Pair` instead to make it clearer.

jhutchison replied:
see above.  (done)

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  @Test
+  public void should_stillGetHScan_givenServerCrashThreadsDoingHScans() {
```
these tests are being scrapped

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,190 @@
+
+  @Test
+  public void should_notLoseDataForConsistentlyPresentFields_givenConcurrentThreadsAddingAndRemovingFields() {
+    final int ITERATION_COUNT = 500;
```
these tests are being scrapped

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,190 @@
+          String field = "new_" + BASE_FIELD + i;
+          jedis3.hset(HASH_KEY, field, "whatever");
+          jedis3.hdel(HASH_KEY, field);
+        }).run();
```
these tests are being scrapped

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+
+  private void multipleHScanAndAssertOnContentOfResultSet(Jedis jedis) {
```
these tests are being scrapped

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -174,13 +211,62 @@ public void givenNonexistentKey_returnsEmptyArray() {
+  }
+
+  @Test
+  public void givenNegativeCursorGreaterThanUnsignedLongCapacity_returnsCursorError() {
```
done

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -237,15 +326,42 @@ public void givenMultipleCounts_returnsAllEntriesWithoutDuplicates() {
+      cursor = result.getCursor();
+    } while (!result.isCompleteIteration());
+
+    assertThat(result.getCursor()).isEqualTo("0");
```
there is a separate test that's been added to verify that hscans don't change the data.  If that doesn't seem like enough let me know.

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +53,91 @@
+  private ConcurrentHashMap<UUID, Long> hScanSnapShotCreationTimes;
+  private ScheduledExecutorService HSCANSnapshotExpirationExecutor = null;
+
+  private static int default_hscan_snapshots_expire_check_frequency =
+      Integer.parseInt(
+          System.getProperty("HSCAN_SNAPSHOTS_EXPIRE_CHECK_FREQUENCY_MILLISECONDS",
+              "30000"));
+
+  private static int default_hscan_snapshots_milliseconds_to_live =
+      Integer.parseInt(
+          System.getProperty("HSCAN_SNAPSHOTS_EXPIRE_CHECK_FREQUENCY_MILLISECONDS",
+              "30000"));
+
```
Integer already has some shortcuts to parse system properties. I'd also suggest the properties be something like `redis.hscan-snapshot-expiry` and `redis.hscan-snapshot-cleanup-interval`. Setting them then becomes: `Integer.getInteger("redis.hscan-snapshot-expiry", 30000)`.

jhutchison replied:
ok-  will change.  thx

jdeppe-pivotal commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +52,68 @@
+    this.hScanSnapShots = new ConcurrentHashMap<>();
+    this.hScanSnapShotCreationTimes = new ConcurrentHashMap<>();
+
+    this.HSCAN_SNAPSHOTS_EXPIRE_CHECK_FREQUENCY_MILLISECONDS =
+        Integer.getInteger("redis.hscan-snapshot-cleanup-interval", 30000);
+
+    this.MINIMUM_MILLISECONDS_FOR_HSCAN_SNAPSHOTS_TO_LIVE =
+        Integer.getInteger("redis.hscan-snapshot-expiry", 30000);
```
System properties like this should be set as `static final` instead of in the constructor since they are only intended to be set once for the life of the JVM. Also related is a comment in the associated test class.

jhutchison replied:
ok- reverted

jdeppe-pivotal commented on geode-redis/src/test/java/org/apache/geode/redis/internal/data/RedisHashTest.java:
```diff
@@ -42,6 +49,15 @@ public static void beforeClass() {
         .getDSFIDSerializer()
         .registerDSFID(DataSerializableFixedID.REDIS_BYTE_ARRAY_WRAPPER, ByteArrayWrapper.class);
+    // setting expiration via system properties
+    System.setProperty("redis.hscan-snapshot-cleanup-interval", "500");
+    System.setProperty("redis.hscan-snapshot-expiry", "500");
```
In general, system properties should be managed with a `RestoreSystemProperties` rule. Also, if possible, unit tests should avoid reliance on system properties. It really turns them into integration tests. I think your previous approach of having a test-only constructor, that took these parameters, was better and I think it would be cleaner to change to that model again.

jhutchison replied:
reverted

nonbinaryprogrammer commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
-      }
+    Integer numberOfIterationsCompleted = indexOfKeys - startCursor;
+
+    Pair resultsPair = new ImmutablePair(numberOfIterationsCompleted, resultList);
```
you could return this directly instead of having this extra assignment

jhutchison replied:
ok

nonbinaryprogrammer commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
+    return keySnapShot;
   }
 
+
```
please put back the formatting here through line 397, as well as your change to the formatting on lines 432&433

jhutchison replied:
not sure where "here" that you're referring to is?

jhutchison replied:
think I got what you were talking about.  Let me know if not

nonbinaryprogrammer commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/HScanExecutor.java:
```diff
@@ -101,20 +105,24 @@ public RedisResponse executeCommand(Command command,
+
       LogService.getLogger().warn(
-          "Could not compile the pattern: '{}' due to the following exception: '{}'. HSCAN will return an empty list.",
+          "Could not compile the pattern: '{}'"
```
please change back this formatting

jhutchison replied:
hmm-  on my machine I don't see this change.  if it comes back let me know and I'll revert it

nonbinaryprogrammer commented on geode-redis/src/main/java/org/apache/geode/redis/internal/netty/ExecutionHandlerContext.java:
```diff
@@ -130,11 +131,15 @@ public ExecutionHandlerContext(Channel channel,
-          response.afterWrite();
-          logResponse(response, channel.remoteAddress().toString(), f.cause());
-        });
+    return channel
```
I don't see any code changes in this block. please revert it

jhutchison replied:
done

nonbinaryprogrammer reviewed (changes requested):
please go through these changes and look for extra line breaks and changes that are formatting-only

sabbey37 commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +53,88 @@
+  private ConcurrentHashMap<UUID, Long> hScanSnapShotCreationTimes;
+  private ScheduledExecutorService HSCANSnapshotExpirationExecutor = null;
+
+  private static int default_hscan_snapshots_expire_check_frequency =
+      Integer.getInteger("redis.hscan-snapshot-cleanup-interval", 30000);
+
+  private static int default_hscan_snapshots_milliseconds_to_live =
+      Integer.getInteger("redis.hscan-snapshot-expiry", 30000);
```
Not sure why we switched from camel case to snake case for variable names?

jhutchison replied:
guess that's a little weird-  changed

sabbey37 reviewed (changes requested):
I haven't finished looking through the entire PR yet, but wanted to post these initial comments and refresh to be up to date with Hale's comments.

jhutchison commented on geode-redis/src/acceptanceTest/java/org/apache/geode/redis/internal/executor/hash/HScanNativeRedisAcceptanceTest.java:
```diff
@@ -27,5 +27,4 @@
   public int getPort() {
     return redis.getPort();
   }
```
done

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HMsetDUnitTest.java:
```diff
@@ -243,6 +243,7 @@ public void shouldDistributeDataAmongMultipleServers_givenMultipleClientsOnSameS
     jedis2B.disconnect();
   }
 
+
```
done

jhutchison commented on geode-redis/src/commonTest/java/org/apache/geode/redis/ConcurrentLoopingThreads.java:
```diff
@@ -63,13 +65,19 @@ private ConcurrentLoopingThreads start(boolean lockstep) {
-        throw new RuntimeException(e);
+    boolean timeOutExceptionThrown;
+    do {
+      timeOutExceptionThrown = false;
```
Yeah, I'll add the time unit to allow it to throw the timeout exception.  Thanks for catching that.

sabbey37 commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/HScanExecutor.java:
```diff
@@ -44,6 +45,9 @@
+
+    final UUID CLIENT_ID = context.getClientUUID();
+
     List<byte[]> commandElems = command.getProcessedCommand();
```
This is my fault... but if we want to match native Redis, the `count` parameter should actually be a long... and only return the `ERR value is not an integer or out of range` message if it exceeds the maximum long capacity (9223372036854775807)... we'll have to change this in all the SCANs (if we decide to).

jhutchison replied:
I think the thought was that a cursor "unit" represents more than a single value in a redis hash (let say 100), and that if we have a max int*100 values in a single hash we would have a different point of failure.

jhutchison replied:
I made a story to look into this parameters for all the scans and document differences.  Going to leave this as an int for now.  Yell at me if you strongly disagree.

sabbey37 replied:
I think it makes sense to leave it as an `int` given the cursor parameter will be an `int`  and max size of a hash is the same as an `int`. Thanks for discussing this with everyone!

sabbey37 commented on geode-redis/src/test/java/org/apache/geode/redis/internal/data/RedisHashTest.java:
```diff
@@ -161,4 +163,109 @@ public void setExpirationTimestamp_stores_delta_that_is_stable() throws IOExcept
+    assertThat(keyList).isNotEmpty();
+
+    FIELDS_AND_VALUES_FOR_HASH.forEach((entry) -> {
+      if (entry.toString().contains("field")) {
+        assertThat(keyList.contains(entry)).isTrue();
+      } else if (entry.toString().contains("value")) {
+        assertThat(keyList.contains(entry)).isFalse();
```
you could do the following:
```
 if (entry.toString().contains("field")) {
        assertThat(keyList).contains(entry);
      } else if (entry.toString().contains("value")) {
        assertThat(keyList).doesNotContain(entry);
      }
 ```
 
 this will result in clearer failure messages.

jhutchison replied:
thanks

jhutchison replied:
changed

sabbey37 reviewed (changes requested):
I always worry that I'm missing something. I may look a bit more tomorrow.

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,299 @@
+  private void crashAlternatingServers(AtomicBoolean keepCrashingServers,
+      AtomicInteger numberOfTimesServersCrashed) {
+
+    int serverToCrashToggle = 3;
```
changed

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,299 @@
+
+
+
+  private void crashAlternatingServers(AtomicBoolean keepCrashingServers,
```
done

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,299 @@
+        }
+      }
+    }
+    keepCrashingServers.set(false);
```
I think if you look closer... this only gets set to false  when control leaves the while with the while (numberOfAssertionsCompleted < 3 || numberOfTimesServersCrashed.get() < 3)  logic.  so it will not be set to false (and stop the vms from crashing). until there have been at least 3 assertions and at least 3 vm crashes.

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -219,13 +298,13 @@ public void givenCount_returnsAllEntriesWithoutDuplicates() {
 
-    assertThat(allEntries).hasSize(3);
-    assertThat(new HashSet<>(allEntries)).isEqualTo(entryMap.entrySet());
+    assertThat(allEntries.size()).isCloseTo(3, Offset.offset(1));
```
see comment at the top of the test - "// note: As per redis docs, count parameter is not guaranteed to be honored."  so it's weird one to test...

sabbey37 replied:
A few thoughts:
- The test name is kinda misleading, since we say `returnsAllEntriesWithoutDuplicates`, but we are admitting the possibility of duplicates.
- You could use lambdas (like you did in another part of the code) to remove any duplicates before the assertion, something like:
```
List<Map.Entry<String, String>> allEntriesWithoutDuplicates = allEntries.stream().distinct().collect(Collectors.toList());
```
- But really none of the above matters because, thinking about it more, I realize that for most of our tests, the data is so small that native Redis is always going to return the entire structure at once (per the `Why SCAN may return all the items of an aggregate data type in a single call?` section of the SCAN documentation as well as the Redis code (see `db.c` file, `scanGenericCommand` method)). So we can actually check that the size and entries match exactly what we expect.

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +441,136 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
-      allEntries.addAll(result.getResult());
-      cursor = result.getCursor();
-    } while (!result.isCompleteIteration());
+    jedis.hdel("colors", "field_3");
```
So hscan promises a few things, essentially boiling down to any change that occurs before an hscan starts will be reflected in the results, but  there are less promises  around that changes that happen during an hscan.   This test is just reflecting the first guarantee-  if a value is delete4d before the hscan, it will not show up ion the hscan (even if it was in the hash before the hscan started).  I agree it seems obvious, but it seems like part of the contract, so it's tested.

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -36,58 +38,86 @@
   @SuppressWarnings("unchecked")
-  public void givenArgumentsAreNotOddAndKeyDoesNotExist_returnsEmptyArray() {
+  public void givenCountArgumentWithoutNumberOnNonExistentKey_returnsEmptyArray() {
+    jedis.hset("a", "b", "1");
```
don't remember when it got added back in, but test  this actually fails for both native redis and our code without the initial set.   I think it's because the key doesn't actually exist w/o it.

sabbey37 replied:
According to the test name, the key **should not exist** before the HSCAN is done. The HSCAN on the line below it is done on a completely different key from the hset (`a` vs. `b`)... so there should be no need to do an `hset` beforehand.  I commented out the hset and ran it on my local machine.... it passed on both native Redis and Geode Redis.

jhutchison replied:
my bad- was looking at this test: givenMatchArgumentWithoutPatternOnExistingKey_returnsSyntaxError().  removed the set

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -237,15 +316,40 @@ public void givenMultipleCounts_returnsAllEntriesWithoutDuplicates() {
+    assertThat(result.getCursor()).isEqualTo("0");
   }
 
   @Test
```
This is just cursor value, but I know what you mean.  I can change them to look for the elements.

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -306,8 +410,9 @@ public void givenMatchAndCount_returnsAllMatchingKeysWithoutDuplicates() {
-    assertThat(allEntries).hasSize(2);
-    assertThat(new HashSet<>(allEntries)).isEqualTo(entryMap.entrySet());
+
+    assertThat(new HashSet<>(allEntries))
+        .containsExactlyInAnyOrderElementsOf(entryMap.entrySet());
```
see above response

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +441,136 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
-  public void givenCursorGreaterThanUnsignedLongCapacity_returnsCursorError() {
-    assertThatThrownBy(() -> jedis.hscan("a", "18446744073709551616"))
-        .hasMessageContaining(ERROR_CURSOR);
+  public void should_retun_not_error_given_non_zero_cursor_on_first_call() {
```
didn't notice.  I can change

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +441,136 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
+  }
+
+  private void multipleHScanAndAssertOnContentOfResultSet(Jedis jedis,
+      final Map<String, String> intialHashData) {
```
done

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,299 @@
+import org.apache.geode.test.junit.rules.GfshCommandRule;
+
+public class HScanDunitTest {
+  @ClassRule
+  public static ClusterStartupRule clusterStartUp = new ClusterStartupRule(4);
```
made this change.  Yeah, there's a fair amount that could be cleaned up/standardized  across the dunit tests when we have the bandwidth.

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
+      int count,
+      BigInteger cursorParameter) {
+
+    int startCursor = cursorParameter.intValue();
```
So @jdeppe-pivotal brought this up in standup the other day.  The decision the team landed on was to make these choices about data types based on our implementation (e.g. what makes sense with java data types in geode) and to guard against data types that we don't support at the command level (send an error back to any client trying to pass a datatype that we don't support internally).  I'll make sure that Hscan follows this convention.  I'll also put in a story to make sure that we're documenting differences between what data types we accept and what native redis accepts.

sabbey37 replied:
That makes sense. Thanks for discussing this!

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -336,48 +441,136 @@ public void givenMultipleCountsAndMatches_returnsEntriesMatchingLastMatchParamet
+    } while (!result.isCompleteIteration());
+
+    assertThat(allEntries.size())
+        .isEqualTo(intialHashData.size());
```
added logic to remove duplicates form the list before assertion.  Good catch

jhutchison replied:
added logic to remove duplicates before asserting on size.  thx

jhutchison commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/AbstractHScanIntegrationTest.java:
```diff
@@ -36,58 +38,86 @@
+
   @Test
-  public void givenNoKeyArgument_returnsWrongNumberOfArgumentsError() {
+  public void givenLessThanTwoArguments_returnsWrongNumberOfArgumentsError() {
```
done

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -46,19 +53,88 @@
+
+  private void expireHScanSnapshots() {
+
+    this.hScanSnapShotCreationTimes.entrySet().forEach(entry -> {
```
o, wow-  that's because I'm not making a copy like I thought I was-  changed to this: 
 List<ByteArrayWrapper> keySnapShot = new ArrayList<>();
    List hashKeysAsList = Arrays.asList(hash.keySet().toArray());
    Collections.copy(keySnapShot, hashKeysAsList);

sabbey37 replied:
I was confused for a second, but then realized the above comment went with the comment that was in the `createKeySnapshot` method. Those changes look good!

jhutchison replied:
yeah sorry, posted responses on the wrong comments a couple of times :(

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
+
+
+  private List<ByteArrayWrapper> createKeySnapShot(UUID clientID) {
+    List<ByteArrayWrapper> keySnapShot = new ArrayList<>();
+    keySnapShot.addAll(hash.keySet());
```
Meant to be copying these.  changed to this (I think I added this comment to the wrong place before) 

List<ByteArrayWrapper> keySnapShot = new ArrayList<>();
    List hashKeysAsList = Arrays.asList(hash.keySet().toArray());
    Collections.copy(keySnapShot, hashKeysAsList);

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,241 @@
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  @ClassRule
+  public static GfshCommandRule gfsh = new GfshCommandRule();
```
yeah, my bad-  I didn't notice the rule was still around.  will remove.

jhutchison commented on geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HScanDunitTest.java:
```diff
@@ -0,0 +1,241 @@
+    server3.stop();
+  }
+
+  @Test
+  public void should_allow_hscan_iteration_to_complete_successfully_given_server_crashes_during_iteration()
```
done

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/HScanExecutor.java:
```diff
@@ -44,26 +45,29 @@
 
     try {
-      cursor = new BigInteger(cursorString).abs();
+      cursor = Integer.valueOf(cursorString);
```
changed

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/HScanExecutor.java:
```diff
@@ -44,26 +45,29 @@
     }
 
-    if (cursor.compareTo(UNSIGNED_LONG_CAPACITY) > 0) {
+    if (cursor > Integer.MAX_VALUE) {
       return RedisResponse.error(ERROR_CURSOR);
     }
```
makes sense.  thx

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/executor/CommandFunction.java:
```diff
@@ -257,11 +258,13 @@ protected Object compute(ByteArrayWrapper key, Object[] args) {
         int count = (int) args[2];
-        BigInteger cursor = (BigInteger) args[3];
-        return hashCommands.hscan(key, pattern, count, cursor);
+        int cursor = Integer.valueOf(args[3].toString());
```
changed to  int cursor = Integer.parseInt((args[3].toString())); not sure if that's what you mean?  
(I can't just cast it because it's a string).

sabbey37 replied:
`args` are an array of Objects (`Object[]`), not strings.  `args[3]` can be cast in the same way that we cast `args[2]` for the `count` variable on the line above it and throughout the `CommandFunction` class:
```
int cursor = (int) args[3];
```

jhutchison replied:
geez, I guess you're right-  looked at that too quickly.  thanks

jhutchison commented on geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java:
```diff
@@ -203,47 +280,118 @@ public int hstrlen(ByteArrayWrapper field) {
+      int count,
+      int cursorParameter) {
+
+    int startCursor = cursorParameter;
```
yeah, Now that's it's an int everywhere this isn't being used.  removed.  Thx

jhutchison commented:
@sabbey (for some reason this particular comment won't let me reply on it )

regarding the test givenCount_returnsAllEntriesWithoutDuplicates() mentioned above- that test seemed not super valuable like it was (testing that the entire set was returned), so I changed it a little and moved it to the (concrete) HScanIntegrationTest (since our implementation doesn't really match native Redis' exactly. ).

sabbey37 commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/HScanIntegrationTest.java:
```diff
@@ -37,22 +44,58 @@ public int getPort() {
+
   @Test
-  public void givenDifferentCursorThanSpecifiedByPreviousHscan_returnsAllEntries() {
+  public void givenCount_shouldReturnsExpectedNumberOfEntries() {
```
If you end up making other changes, the grammar here could be corrected (either changed to `givenCount_shouldReturnExpectedNumberOfEntries` or (more consistent with the tests above it) `givenCount_returnsExpectedNumberOfEntries`.

jhutchison replied:
oops.  thanks

sabbey37 commented on geode-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/hash/HScanIntegrationTest.java:
```diff
@@ -37,22 +44,58 @@ public int getPort() {
+
+
   @Test
-  public void givenDifferentCursorThanSpecifiedByPreviousHscan_returnsAllEntries() {
+  public void givenCount_shouldReturnsExpectedNumberOfEntries() {
     Map<String, String> entryMap = new HashMap<>();
-    for (int i = 0; i < 10; i++) {
-      entryMap.put(String.valueOf(i), String.valueOf(i));
-    }
-    jedis.hmset("a", entryMap);
+    entryMap.put("1", "yellow");
+    entryMap.put("2", "green");
+    entryMap.put("3", "orange");
+    jedis.hmset("colors", entryMap);
+
+    int COUNT_PARAM = 2;
 
     ScanParams scanParams = new ScanParams();
-    scanParams.count(5);
-    ScanResult<Map.Entry<String, String>> result = jedis.hscan("a", "0", scanParams);
-    assertThat(result.isCompleteIteration()).isFalse();
+    scanParams.count(COUNT_PARAM);
+    ScanResult<Map.Entry<String, String>> result;
+
+    List<Map.Entry<String, String>> allEntries = new ArrayList<>();
+    String cursor = "0";
 
-    result = jedis.hscan("a", "100");
+    result = jedis.hscan("colors", cursor, scanParams);
+    allEntries.addAll(result.getResult());
 
-    assertThat(result.getResult()).hasSize(10);
-    assertThat(new HashSet<>(result.getResult())).isEqualTo(entryMap.entrySet());
+    List<Map.Entry<String, String>> allDistinctEntries =
+        allEntries
+            .stream()
+            .distinct()
+            .collect(Collectors.toList());
+
+    assertThat(allDistinctEntries.size()).isEqualTo(COUNT_PARAM);
```
I like this test. Thanks for thinking of it!  Because we are making sure that we return the same number of entries as the `COUNT_PARAM`, it doesn't make sense to take out duplicates (we should return the same number of entries as the `COUNT_PARAM`, including duplicates). In that case, lines 90-96 could be removed and we could just assert the following:
`assertThat(result.getResult().size()).isEqualTo(COUNT_PARAM)`

jhutchison replied:
Yeah, I'm hesitant to read too much into the exactness of what the count parameter does.  as per docs (as I'm sure you've read) " Basically with COUNT the user specified the amount of work that should be done at every call in order to retrieve elements from the collection. This is just a hint for the implementation, however generally speaking this is what you could expect most of the times from the implementation."   I'll remove the removal of the duplicates, though

sabbey37 replied:
If this were running with native Redis, it would return all entries because it is such a small amount of data.  In that case, removing duplicates would not have any effect (all elements would be returned at once without duplicates, so the amount of elements returned would always be equal to the size of the complete entry map rather than the count parameter [read `Why SCAN may return all the items of an aggregate data type in a single call?` in the [SCAN documentation](https://redis.io/commands/scan) or look at the Redis code for more on this]).  

However, this test is only run with Geode's Redis compatibility API.  I thought we moved it to `HScanIntegrationTest.java` specifically so our method of returning with a count param (which only returns the exact number of elements specified by count) could be tested.

sabbey37 reviewed (approved):
Thank you for making all those changes!
